d2themes/element.go

```diff
@@ -227,3 +227,13 @@ func (el *ThemableElement) Render() string {
 	}
 	return out
 }
+
+func calculateAxisRadius(borderRadius float64, sideLength float64) float64 {
```

Comment:
I've found calculating a value to use for `rx`/`ry` instead of passing it a percentage value makes it easier to render as is done in CSS. This way we avoid the case where having multiple rectangles with percentage rx/ry will render differently 

> Exemple of what was wrong with % values in `rx`, 5% border-radius behaved differently if it was rendered with other shapes.
<img width="400px" alt="example2" src="https://user-images.githubusercontent.com/15009534/224089772-b5101095-1357-4060-aaa2-226591cf452e.png" />

<img width="400px" alt="example4" src="https://user-images.githubusercontent.com/15009534/224089786-5ba28b39-4552-4e9e-a055-dea6e1a753c4.png" />


The following exemples are rendered with this new function, calculating the border radius to set it with a number, not a percentage

> Multiple shape with border-radius 10%
<img src="https://user-images.githubusercontent.com/15009534/225982272-486184e2-3b24-4626-9883-eb285b409247.png" width="400px"/>

> Html/css with border-radius 10%, it is the same shape 👍 
<img src="https://user-images.githubusercontent.com/15009534/225982276-56552748-0a66-46a3-9d84-75af1cf6760e.png" width="200px"/>

> A shape with border-radius 35% and the same shape using html/css with border-radius 35%
<img src="https://user-images.githubusercontent.com/15009534/225982281-f0df20fa-9a87-4a35-b29a-77084bfeef5e.png" width="200px"/>
<img src="https://user-images.githubusercontent.com/15009534/225982283-7ce1752d-9792-4610-9885-33d8c10b5e33.png" width="200px"/>

> Border-radius 50%
<img src="https://user-images.githubusercontent.com/15009534/225984408-a0080c9a-7fcf-46f8-8c4d-a4f7e5381c74.png" width="200px"/>


## Questions ❓ 
* Do you think this is a good way approach this feature ?

If yes -> What bothers me a bit is that `calculateAxisRadius` now depends on either `Height` or `Width` when dealing with percentages values, and I did not see any other example of a computed value like this in this file. So we need to check if the "side length" value equality to `math.MaxFloat64` one more time here.
The code itself could be refactored to be more clear (maybe something like `out += el.getRx()` instead of `if el.Rx != math.MaxFloat64 {out += fmt.Sprintf(' rx="%f"', calculateAxisRadius(el.Rx, el.Width))}` to handle necessary check on Rx and the conditional check on Width in the same function `el.getRx()`, instead of doing validity check outside and inside the "calculate" function)


* Does this issue includes pill shapes ?

Another point of attention is that with only one border-radius percentage value, it is not possible to have pill shape like this.
<img src="https://user-images.githubusercontent.com/15009534/225984406-5f2612fa-5bf4-47ec-926e-2bf13e71c26a.png" width="200px"/>

To achieve that there is multiple solutions, but none with % in rx or  % border-radius in CSS. For exemple one solution is to set `rx` to 50% of the height of the rectangle (and give either no value or the same value to `ry`).

So the current implementation is good at replicating CSS when given a % value in `border-radius`, but cannot handle pill shape (CSS can't either with a  % `border-radius`).
Does this PR have to handle pill shape or do you think this is another issue ? If it has to, how do you think it should be implemented ? Add a `pill: true` property to the rectangle shape or create a new pill shape (that could be derived from rectangle shape in code if necessary)?

Reply:
apologies for the late response here. 

> This way we avoid the case where having multiple rectangles with percentage rx/ry will render differently

> What bothers me a bit is that calculateAxisRadius now depends on either Height or Width when dealing with percentages value

I don't understand, isn't that the point of percentages? To adapt to dimensions?

> if it was rendered with other shapes.

i also don't understand. what shapes are you using? are those all rectangles or are some ovals?

> Does this issue includes pill shapes ?

Firstly, really appreciate you thinking of motivating use cases like this. I wonder if this would work https://stackoverflow.com/a/18795153/20901387 . Border radius currently has a ceiling, but we could remove that for this use case (https://github.com/terrastruct/d2/issues/948).

Reply:
>This way we avoid the case where having multiple rectangles with percentage rx/ry will render differently

I wanted to highlight the issue that two rectangles with the same percentage `rx` did not render the same way if rendered in svg of different size. Eg. In the first two images of this thread, https://user-images.githubusercontent.com/15009534/224089772-b5101095-1357-4060-aaa2-226591cf452e.png https://user-images.githubusercontent.com/15009534/224089786-5ba28b39-4552-4e9e-a055-dea6e1a753c4.png, the two rectangle "ElementWithALongerName0,05" have a border-radius of 5% (rx) but are not rendered the same. That is why I explored the solution of computing a "number" value for rx/ry instead of simply passing a percentage value.

>I don't understand, isn't that the point of percentages? To adapt to dimensions?

Sure I agree 😄. I think this solution can work. I'm not familiar enough with the project to know if we will always have a `height` & `width` values when passing a `border-radius`, if we don't then it's impossible in those cases to compute the value and we would have to pass the raw percentage value, or nothing. What do you think about it, would it be acceptable to "compute" a rx/ry when given a % border-radius? Or would we prefer passing the raw % value to our shape and find another solution to make sure the way they render is consistant (solve the issue of the "ElementWithALongerName0,05" example)?

> I also don't understand. what shapes are you using? are those all rectangles or are some ovals?

What I meant by "if it was rendered with other shapes." is that "ElementWithALongerName0,05" is not rendered the same in the two exemple because the second example has other shapes (rect) in it, that is somehow causing a different render of its shape even if they have the same border-radius % value. I wanted to illustrate the issue I had with % rx/ry values.
That is why I tried to compute a non-percentage rx/ry value from a % border-radius value, that solution is illustrated with the other examples.
Every example you see are svg rect shapes with different border-radius values. The orange one are html/css with border-radius value to compare the results with.

> Firstly, really appreciate you thinking of motivating use cases like this. I wonder if this would work https://stackoverflow.com/a/18795153/20901387 . Border radius currently has a ceiling, but we could remove that for this use case (https://github.com/terrastruct/d2/issues/948).

I tried to do this (passing a huge value to rx/ry inside a rect), but that just renders rectangles as oval (like having %50 border-radius). The best solution I found for making a pill is to set `rx` to half of the `height` value. So we could always compute the `rx` value using the given percentage and the smallest value between `width` and `height`, and 50% would be a pill shape, but that would make us diverge a lot from CSS.
So if you consider the pill shape to be part of this issue, we have to find a solution (maybe creating a pill shape, of maybe a props to rect shape, I'm not sure which is the best). But if you don't, this PR can focus on replicating the CSS `border-radius` with a percentage value

Reply:
> I wanted to highlight the issue that two rectangles with the same percentage rx did not render the same way if rendered in svg of different size

wait are you saying that it's dependent of the whole diagram size? not just the object size?

can you copy the .d2 code for those examples? or just include them as tests in `todo_test` and generate the svgs

Reply:
Exactly, that's what I have found but I might miss something as I'm kinda new on this kind of subject.
Here is a gist to illustrate it https://gist.github.com/Poivey/d965e66456925e41c19a3f179b18a0ee

As you can see here, both `percent-5percent-*.svg` files are containing a `rect` with `rx` and `ry` with `5%` value. But `ElementWith5PercentBorder` is not rendered the same way in the two file, when it is the only shape in the svg ([percentage-5percent-alone.svg](https://gist.github.com/Poivey/d965e66456925e41c19a3f179b18a0ee#file-percentage-5percent-alone-svg)) its border radius is rendered a different way than when it is rendered with other elements in the svg ([percentage-5percent-with-other-elements.svg](https://gist.github.com/Poivey/d965e66456925e41c19a3f179b18a0ee#file-percentage-5percent-with-other-elements-svg)).

I also added two file [numeric-5percent-with-other-elements.svg](https://gist.github.com/Poivey/d965e66456925e41c19a3f179b18a0ee#file-numeric-5percent-with-other-elements-svg) and [numeric-5percent-alone.svg](https://gist.github.com/Poivey/d965e66456925e41c19a3f179b18a0ee#file-numeric-5percent-alone-svg) that show that when computing a numeric value (and not a %) to set `rx` and  `ry` with, the `rect` is rendered the same way every time AND matches the shape of a css element with `border-radius: 50%` -> This is the current state of this PR

d2 souce is in the gist, it's just this. `AnotherElementX` are commented when rendering the rectangle with border-radius alone.
```
ElementWith5PercentBorder.style.border-radius: 0.05
AnotherElement1
AnotherElement2
AnotherElement3
AnotherElement4
```